sdk/src/androidTest/java/com/mapbox/maps/ViewAnnotationTest.kt

```diff
@@ -0,0 +1,744 @@
+import com.mapbox.maps.extension.style.sources.addSource
+import com.mapbox.maps.extension.style.sources.generated.geoJsonSource
+import com.mapbox.maps.test.R
+import com.mapbox.maps.viewannotation.ViewAnnotationManager
+import com.mapbox.maps.viewannotation.viewAnnotationOptions
+import org.hamcrest.MatcherAssert
+import org.hamcrest.Matchers
+import org.junit.*
+import org.junit.Assert.*
+import org.junit.runner.RunWith
+import java.util.concurrent.CountDownLatch
+
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+class ViewAnnotationTest {
+  private lateinit var mapboxMap: MapboxMap
+  private lateinit var mapView: MapView
+  private lateinit var viewAnnotationManager: ViewAnnotationManager
+  private lateinit var mainHandler: Handler
+
+  @get:Rule
+  var rule = ActivityScenarioRule(EmptyActivity::class.java)
+
+  @Before
+  fun before() {
+    val latch = CountDownLatch(2)
+    rule.scenario.onActivity {
+      it.runOnUiThread {
+        val context = InstrumentationRegistry.getInstrumentation().targetContext
+        mapView = MapView(context)
+        mapView.id = R.id.mapView
+        it.setContentView(mapView)
+
+        mainHandler = Handler(Looper.getMainLooper())
+        viewAnnotationManager = mapView.viewAnnotationManager
+        mapboxMap = mapView.getMapboxMap().apply {
+          loadStyleUri(
+            Style.MAPBOX_STREETS
+          ) {
+            latch.countDown()
+          }
+          setCamera(
+            CameraOptions.Builder()
+              .center(CAMERA_CENTER)
+              .zoom(CAMERA_ZOOM)
+              .build()
+          )
+          addOnMapIdleListener {
+            latch.countDown()
+          }
+        }
+        mapView.onStart()
+      }
+    }
+    latch.throwExceptionOnTimeoutMs(CONFIGURE_TIMEOUT_MS)
+  }
+
+  @After
+  fun tearDown() {
+    val latch = CountDownLatch(1)
+    rule.scenario.onActivity {
+      it.runOnUiThread {
+        mapView.onStop()
+        mapView.onDestroy()
+        mainHandler.removeCallbacksAndMessages(null)
+        latch.countDown()
+      }
+    }
+    latch.throwExceptionOnTimeoutMs(CONFIGURE_TIMEOUT_MS)
+  }
+
+  // checking some use-cases when adding view annotation
+
+  @Test
+  fun addViewAnnotationNoOptions() {
+    val latch = CountDownLatch(1)
+    mainHandler.post {
+      val view = viewAnnotationManager.addViewAnnotation(
+        resId = R.layout.view_annotation,
+        options = viewAnnotationOptions {
+          geometry(CAMERA_CENTER)
+        }
+      )
+      mainHandler.postDelayed(
+        {
+          assertTrue(mapView.hasChildView(view))
+          assertEquals(mapboxMap.pixelForCoordinate(CAMERA_CENTER).x - view.width / 2.0, view.translationX.toDouble(), ADMISSIBLE_ERROR_PX)
+          assertEquals(mapboxMap.pixelForCoordinate(CAMERA_CENTER).y - view.height / 2.0, view.translationY.toDouble(), ADMISSIBLE_ERROR_PX)
+          latch.countDown()
+        },
+        VIEW_PLACEMENT_DELAY_MS
```

Comment:
This is for letting gl native time to process added view?

Reply:
exactly, we have internal listener for that but there's no guarantee it will be triggered once, it might trigger several times till exact placement will arrive.
in general core listens to internal Observer API event so we can't easily hook in and we have a delay here